src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -1286,159 +1213,29 @@ internal void SetAuthenticationHeaders(HttpListenerContext context)
             return result;
         }
 
-        private bool CheckSpn(NTAuthentication context, bool isSecureConnection, ExtendedProtectionPolicy policy)
+        private ExtendedProtectionPolicy GetAuthenticationExtendedProtectionPolicy(ExtendedProtectionPolicy policy)
         {
-            // Kerberos does SPN check already in ASC
-            if (context.IsKerberos)
+            if (policy.PolicyEnforcement == PolicyEnforcement.Never ||
+                policy.CustomServiceNames != null)
             {
-                if (NetEventSource.Log.IsEnabled())
-                {
-                    NetEventSource.Info(this, SR.net_log_listener_no_spn_kerberos);
-                }
-                return true;
-            }
-
-            // Don't check the SPN if Extended Protection is off or we already checked the CBT
-            if (policy.PolicyEnforcement == PolicyEnforcement.Never)
-            {
-                if (NetEventSource.Log.IsEnabled())
-                {
-                    NetEventSource.Info(this, SR.net_log_listener_no_spn_disabled);
-                }
-                return true;
+                return policy;
             }
 
-            if (ScenarioChecksChannelBinding(isSecureConnection, policy.ProtectionScenario))
+            if (_defaultServiceNames.ServiceNames.Count == 0)
             {
-                if (NetEventSource.Log.IsEnabled())
-                {
-                    NetEventSource.Info(this, SR.net_log_listener_no_spn_cbt);
-                }
-                return true;
+                throw new InvalidOperationException(SR.net_listener_no_spns);
             }
 
-            string? clientSpn = context.ClientSpecifiedSpn;
+            // TODO: Do we want to allow http/localhost?
```

Comment:
There's a change in observable behavior. Previously the `http/localhost` SPN was always allowed on NTLM connections. Now it's only allowed when explicitly specified in the `ExtendedProtectionPolicy` policy as custom service name, or if custom service names are not specified and `http/localhost` is included in the default SPNs generated by `HttpListener.Prefixes.Add` call.

(This doesn't affect Kerberos in any way.)

Reply:
If there is no good reason (like security) I would prefer to keep the old behavior. As @stephentoub  mentioned this is old code with little coverage and used mostly for legacy reasons. So I would feel compat and preserving existing functionality is important - perhaps even more than let say HttpClient.

Reply:
The reason why it is changed is to match how SPN is treated in NegotiateStream/NegotiateAuthentication. Adding an extra prefix is possible but it means rewriting the allowed SPN lists and explicitly adding it.

Reply:
The change turned out to be quite small in the end, so I added "HTTP/localhost" to the list.

I assume it makes sense since lot of the `HttpListener` examples use a code like this:
```csharp
        HttpListener listener = new HttpListener();
        listener.Prefixes.Add("http://*:8080/");
```
In this case the default SPN list is populated with one entry `HTTP/<computer name>`, which in my case would be `HTTP/FilipNavaraR7K`. It seems reasonable to allow `HTTP/localhost` for backwards compatibility.